File: GoSubmarine/day14/extendedPolymerization.go
package main

import (
	"math"
	"strings"

	"github.com/tkrehbiel/adventofcode2021/gosubmarine/common"
)

// https://adventofcode.com/2021/day/14
// Extended Polymerization
/*
This is a Golang translation of my original Kotlin solution.
*/

func main() {
	common.Run(common.PuzzleConfig{
		Puzzle:      new(puzzle),
		Day:         14,
		Title:       "Extended Polymerization",
		PuzzleInput: "day14_input.txt",
		TestInput:   []string{"day14_test.txt"},
	})
}

type puzzle struct {
	initial      string
	instructions map[string]rune
	pairs        map[string]int
	counts       map[rune]int
}

// Common reads the initial puzzle input.
func (p *puzzle) Common(input string) {
	p.instructions = make(map[string]rune)
	common.EnumerateLines(input, func(row int, s string) {
		if row == 0 {
			p.initial = s
		} else {
			if len(s) > 0 {
				parts := strings.Split(s, " -> ")
				p.instructions[parts[0]] = rune(parts[1][0])
			}
		}
	})
}

// Part1 runs the puzzle for 10 iterations
func (p *puzzle) Part1(input string) (result int) {
	p.initMaps()
	return p.run(10)
}

// Part2 runs the puzzle for 40 iterations
func (p *puzzle) Part2(input string) (result int) {
	p.initMaps()
	return p.run(40)
}

// initMaps prepares the letter count and pairs maps
func (p *puzzle) initMaps() {
	p.counts = make(map[rune]int)
	for _, c := range p.initial {
		p.counts[c]++
	}

	p.pairs = make(map[string]int)
	for i := 0; i < len(p.initial)-1; i++ {
		pair := string(p.initial[i]) + string(p.initial[i+1])
		p.pairs[pair]++
	}
}

// run iterates the puzzle a given number of times
func (p *puzzle) run(iterations int) int {
	for i := 0; i < iterations; i++ {
		p.iterate()
	}
	return max(p.counts) - min(p.counts)
}

// iterate runs one iteration of the puzzle
func (p *puzzle) iterate() {
	newPairs := make(map[string]int)
	for pair, count := range p.pairs {
		insert := p.instructions[pair]
		pair1 := string(pair[0]) + string(insert)
		pair2 := string(insert) + string(pair[1])
		p.counts[insert] += count
		newPairs[pair1] += count
		newPairs[pair2] += count
	}
	p.pairs = newPairs
}

// min gets the minimum count in the letter map
func min(counts map[rune]int) int {
	minValue := math.MaxInt64
	for _, v := range counts {
		if v < minValue {
			minValue = v
		}
	}
	return minValue
}

// max gets the maximum count in the letter map
func max(counts map[rune]int) int {
	maxValue := -1
	for _, v := range counts {
		if v > maxValue {
			maxValue = v
		}
	}
	return maxValue
}

File: KotlinSubmarine/src/main/kotlin/day14/ExtendedPolymerization.kt
package day14

import Puzzle
import java.io.File

fun main() {
    ExtendedPolymerization().run()
}

// https://adventofcode.com/2021/day/14
/*
Another annoying one where brute force on part 1 runs out of memory on part 2.
The clue is in the offhand “polymer grows quickly” remark. I knew my first
solution wasn't going to work for part 2 but you’re scored on how fast you get
an answer, not how elegant or efficient the code is. So I got the star for part 1
fairly quickly, then I had to rewrite it from scratch for part 2.

I didn't think a map of letter pairs would work so I went off on several false starts,
but it turned out it worked fine in the end.

Just for fun I re-wrote the solution in Golang, and it looked cleaner than Kotlin.
*/

class ExtendedPolymerization : Puzzle(14, "Extended Polymerization") {
    override val puzzleInput = "day14_input.txt"
    override val testInput = arrayOf("day14_test.txt")
    override val label1 = "after 10 steps"
    override val label2 = "after 40 steps"

    private var initial: String = ""
    private val instructions = mutableMapOf<String, Char>()

    override fun common(input: String) {
        super.common(input)

        val lines = File(input).readLines()
        initial = lines[0]

        // Assumes instructions begin on the third line
        instructions.clear()
        for (i in 2 until lines.size) {
            val parts = lines[i].split("->")
            val pair = parts[0].trim()
            val insertion = parts[1].trim()[0]
            instructions[pair] = insertion
        }
    }

    override fun part1(input: String): Long {
        return polymerization(initial, instructions, 10)
    }

    override fun part2(input: String): Long {
        // Technically could let part 1 run for 30 more iterations, but we start over.
        return polymerization(initial, instructions, 40)
    }
}

private fun polymerization(initial: String,
                           instructions: MutableMap<String, Char>,
                           iterations: Int): Long {
    val counts = initCounts(initial)
    var pairs = initPairs(initial)
    for (i in 0 until iterations) {
        pairs = process(pairs, counts, instructions)
    }
    return max(counts) - min(counts)
}

private fun process(map: MutableMap<String, Long>,
                    counts: MutableMap<Char, Long>,
                    instructions: MutableMap<String, Char>): MutableMap<String, Long> {
    val newMap = mutableMapOf<String, Long>()
    for (pair in map.keys) {
        val insert = instructions[pair]!!
        val pair1: String = pair[0].toString() + insert
        val pair2: String = insert.toString() + pair[1]
        val count = map[pair]!!
        add(counts, insert, count)
        add(newMap, pair1, count)
        add(newMap, pair2, count)
    }
    return newMap
}

private fun initCounts(initial: String): MutableMap<Char, Long> {
    val counts = mutableMapOf<Char, Long>()
    counts.clear()
    for (c in initial) increment(counts, c)
    return counts
}

private fun initPairs(initial: String): MutableMap<String, Long> {
    val pairs = mutableMapOf<String, Long>()
    for (i in 0 until initial.length - 1) {
        val pair = initial.substring(i, i + 2)
        add(pairs, pair, 1)
    }
    return pairs
}

// I can't believe I have to write a function for this
// Kotlin is neat but stuff like this seems ridiculous
// Based on https://www.techiedelight.com/increment-value-map-kotlin/
private fun <K> add(map: MutableMap<K, Long>, key: K, n: Long) {
    val value = if (map.containsKey(key)) map[key] else 0
    map[key] = value!! + n
}

private fun <K> increment(map: MutableMap<K, Long>, key: K) {
    add(map, key, 1)
}

private fun min(map: MutableMap<Char, Long>): Long {
    var min = Long.MAX_VALUE
    for (c in map.keys) {
        if (map[c]!! < min) min = map[c]!!
    }
    return min
}

private fun max(map: MutableMap<Char, Long>): Long {
    var max = Long.MIN_VALUE
    for (c in map.keys) {
        if (map[c]!! > max) max = map[c]!!
    }
    return max
}

/*
Old brute force code for posterity:

private fun bruteForce(initial: String, instructions: MutableMap<String, Char>, iterations: Long) {
    // I have a bad feeling they're going to set the iterations
    // to a value that's too high to process like this in part 2
    // Yeah.. in part two it goes from 10 to 40 iterations.
    // And yeah, it becomes impossible to brute force it.
    var s = initial
    for (i in 0 until iterations) {
        s = bruteProcess(s, instructions)
        //println("length after ${i+1}: ${s.length}")
        //println(s)
    }
    //for (c in counts.keys) {
    //    println("number of $c: ${counts[c]}")
    //}
    var result = max(counts) - min(counts)
    println("first attempt result: $result")
}

private fun bruteProcess(s: String, instructions: MutableMap<String, Char>): String {
    var builder = StringBuilder(s.length * 2)
    var final = '0'
    for (i in 0 until s.length - 1) {
        val pair = s.substring(i, i + 2)
        val insert: Char = instructions[pair]!!
        //println("Insert $insert between $pair")
        builder.append(pair[0])
        builder.append(insert)
        increment(counts, insert)
        final = pair[1]
    }
    builder.append(final)
    return builder.toString()
}
*/
